### PR TITLE
refactor - granting permissions

### PR DESCRIPTION
**Pull request by sravanmedarapu**

### CODE REVIEW

File: lib/tools/adb-commands.js [hunk 1]
```diff
@@ -69,11 +69,30 @@ methods.clear = async function (pkg) {
...
+methods.grantAllPermissions = async function (pkg, apk) {
+    let apiLevel = await this.getApiLevel();
+    let targetSdk = null;
+    try {
+      targetSdk = await this.targetSdkVersionFromManifest(apk);
+    } catch (e) {
+      //avoiding logging error stack, as calling library function would have logged
+      log.warn(`Ran into problem getting target SDK version; ignoring...`);
+    }
+    if (apiLevel >= 23 && targetSdk >= 23) {
```
Comment (jlipps) [inline lib/tools/adb-commands.js]: doesn't this need to have `(!targetSdk || targetSdk >= 23)`?

Comment (sravanmedarapu) [inline lib/tools/adb-commands.js]: Tried with different combinations with and with out having target sdk specified in app, grant permission respected only when application defined with targetSdk.

Official Google documentation also mention same
https://developer.android.com/training/permissions/requesting.html

File: lib/tools/adb-commands.js [hunk 2]
```diff
@@ -73,10 +73,14 @@ methods.grantAllPermissions = async function (pkg, apk) {
...
-      targetSdk = await this.targetSdkVersionFromManifest(apk);
+      if (apk === undefined || apk === null) {
+        targetSdk = await this.targetSdkVersionUsingPKG(pkg);
+      } else {
+        targetSdk = await this.targetSdkVersionFromManifest(apk);
+      }
     } catch (e) {
       //avoiding logging error stack, as calling library function would have logged
-      log.warn(`Ran into problem getting target SDK version; ignoring...`);
+      log.warn(`Ran into problem getting target SDK version; ignoring...`+e.message);
```
Comment (jlipps) [inline lib/tools/adb-commands.js]:
```
log.warn(`Ran into problem getting target SDK version; ignoring... ${e.message}`);
```

File: lib/tools/adb-commands.js [hunk 3]
```diff
@@ -73,10 +73,14 @@ methods.grantAllPermissions = async function (pkg, apk) {
     let apiLevel = await this.getApiLevel();
     let targetSdk = null;
     try {
-      targetSdk = await this.targetSdkVersionFromManifest(apk);
+      if (apk === undefined || apk === null) {
```
Comment (jlipps) [inline lib/tools/adb-commands.js]:
```js
if (!apk) {
```

File: lib/tools/android-manifest.js [hunk 4]
```diff
@@ -105,6 +105,17 @@ manifestMethods.targetSdkVersionFromManifest = async function (localApk) {
...
 
+manifestMethods.targetSdkVersionUsingPKG = async function (pkg) {
+  let stdout =  await this.shell(['dumpsys', 'package', pkg]);
+  let targetSdkVersion = new RegExp(/targetSdk=([^\s\s]+)/g).exec(stdout);
+  if (targetSdkVersion && targetSdkVersion.length >= 2) {
+    targetSdkVersion = targetSdkVersion[1];
+  } else {
+    targetSdkVersion = null;
+  }
+  return parseInt(targetSdkVersion, 10);
```
Comment (jlipps) [inline lib/tools/android-manifest.js]: `parseInt(null, 10)` returns `NaN` which is probably not what we mean. I'd either return `null` explicitly or return 0.

File: lib/tools/adb-commands.js (hunk 1 above)
Comment (jlipps) [inline lib/tools/adb-commands.js]: ok makes sense

Comment (mykola-mokhnach) [inline lib/tools/adb-commands.js]: shouldn't this look like 

```
apiLevel >= 23 && targetSdk && targetSdk >= 23
```

in such case? What is the result of null >= 23 comparison?

Comment (jlipps) [inline lib/tools/adb-commands.js]: `null >= 23` is `false` so we should be ok

**Comment by jlipps** [comment]
@sravanmedarapu do functional and unit tests pass with these changes?

**Comment by jlipps** [comment]
aha, I had the feeling we might need a test update :-)

**Comment by jlipps** [comment]
@sravanmedarapu if you've run all the unit and functional tests locally, i'm fine to not wait for travis.

**Comment by sravanmedarapu** [comment]
travis execution from my forked [repo](https://travis-ci.org/sravanmedarapu/appium-adb/builds/182727085) passing  and functional tests are also passing .
I think we can proceed.

**Comment by jlipps** [comment]
ok, publishing in 2.9.2

**Comment by sravanmedarapu** [comment]
will update in other drivers.

**Comment by sravanmedarapu** [comment]
#https://github.com/appium/appium/issues/7493

**Pull request merged**